### PR TITLE
Added genesis config option for elliptic curves.

### PR DESCRIPTION
**Pull request by daniel-iobuilders**
Signed-off-by: Daniel Lehrner <daniel@io.builders>

This pull request add a new optional configuration option to the genesis file. The optional `ecCurve` option can be used to define alternative elliptic curves. This PR currently only supports SECP256K1, no other one. Once this PR is merged, #2008 will be updated, to use this configuration option to select SECP256R1.

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 1]
```diff
@@ -2628,4 +2630,23 @@ public int getDatabaseVersion() {
...
     }
   }
+
+  private void instantiateSignatureAlgorithmFactory() {
+    if (SignatureAlgorithmFactory.isInstanceSet()) {
+      return;
+    }
+
+    if (genesisFile == null) {
+      SignatureAlgorithmFactory.setInstance(SignatureAlgorithmType.create(Optional.empty()));
```
Comment (lucassaldanha) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2640]: Will that use the default key algorith? We could simplify this code a bit adding a method to the SignatureAlgorithmFactory like:
```
SignatureAlgorithmFactory.setDefault();
```
Makes it easier to read.

File: crypto/src/main/java/org/hyperledger/besu/crypto/SignatureAlgorithmType.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+package org.hyperledger.besu.crypto;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Supplier;
+
+public class SignatureAlgorithmType {
+
+  private static final Map<String, Supplier<SignatureAlgorithm>> TYPES =
```
Comment (lucassaldanha) [inline crypto/src/main/java/org/hyperledger/besu/crypto/SignatureAlgorithmType.java:24]: Could we rename this to `SUPPORTED_ALGORITHMS`

File: crypto/src/test/java/org/hyperledger/besu/crypto/SignatureAlgorithmFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
+/*
+ *
+ *  * Copyright ConsenSys AG.
```
Comment (lucassaldanha) [inline crypto/src/test/java/org/hyperledger/besu/crypto/SignatureAlgorithmFactoryTest.java:3]: I think you've got too many `*` in each line :)

File: crypto/src/test/java/org/hyperledger/besu/crypto/SignatureAlgorithmTypeTest.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
+/*
+ *
+ *  * Copyright ConsenSys AG.
+ *  *
+ *  * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
```
Comment (lucassaldanha) [inline crypto/src/test/java/org/hyperledger/besu/crypto/SignatureAlgorithmTypeTest.java:5]: Same, the license format seems to be wrong

**Comment by lucassaldanha** [comment]
I think it would be good to have an INFO log message when the key is set to anything different from the default. We probably don't want to log anything if the default is used, but for other types.

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 1 above)
Comment (lucassaldanha) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2640]: Actually, do we even need to set anything? SignatureAlgorithFactory.getInstance() has a default value, hasn't it?

Comment (daniel-iobuilders) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: You are right, we don't need to set anything here. I have replaced  as well the type of `Optional<String>` simply by `String`. That makes the code a bit easier to read.

File: crypto/src/test/java/org/hyperledger/besu/crypto/SignatureAlgorithmFactoryTest.java (hunk 3 above)
Comment (daniel-iobuilders) [reply crypto/src/test/java/org/hyperledger/besu/crypto/SignatureAlgorithmFactoryTest.java]: I've fixed it

File: crypto/src/test/java/org/hyperledger/besu/crypto/SignatureAlgorithmTypeTest.java (hunk 4 above)
Comment (daniel-iobuilders) [reply crypto/src/test/java/org/hyperledger/besu/crypto/SignatureAlgorithmTypeTest.java]: Thanks. I've fixed it

File: crypto/src/main/java/org/hyperledger/besu/crypto/SignatureAlgorithmType.java (hunk 2 above)
Comment (daniel-iobuilders) [reply crypto/src/main/java/org/hyperledger/besu/crypto/SignatureAlgorithmType.java]: Yes! That name is so much better.

**Comment by daniel-iobuilders** [comment]
> I think it would be good to have an INFO log message when the key is set to anything different from the default. We probably don't want to log anything if the default is used, but for other types.

I will do this in #2008 . In this PR here only the default can be set, so there is not yet a need for the log message.

**Review by lucassaldanha — APPROVED** [review]

**Comment by matkt** [comment]
When I start a node with the commit of this PR on master I have an issue 

I tried with the commit just before and it's was ok


> {"timestamp":"2021-03-22T10:07:35,093","level":"WARN","thread":"vertx-blocked-thread-checker","class":"BlockedThreadChecker","message":"Thread Thread[vert.x-worker-thread-6,5,main]=Thread[vert.x-worker-thread-6,5,main] has been blocked for 62907 ms, time limit is 60000 ms","throwable":" io.vertx.core.VertxException: Thread blocked\n\tat java.base@11.0.10/java.util.HashMap.putVal(HashMap.java:637)\n\tat java.base@11.0.10/java.util.HashMap.put(HashMap.java:607)\n\tat java.base@11.0.10/java.security.Provider.implPut(Provider.java:1064)\n\tat java.base@11.0.10/java.security.Provider.put(Provider.java:511)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.addAlgorithm(BouncyCastleProvider.java:292)\n\tat app//org.bouncycastle.jcajce.provider.asymmetric.DH$Mappings.configure(Unknown Source)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.loadAlgorithms(BouncyCastleProvider.java:245)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.setup(BouncyCastleProvider.java:171)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.access$000(BouncyCastleProvider.java:58)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider$1.run(BouncyCastleProvider.java:153)\n\tat java.base@11.0.10/java.security.AccessController.doPrivileged(Native Method)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.<init>(BouncyCastleProvider.java:149)\n\tat app//org.hyperledger.besu.crypto.SECP256K1.<init>(SECP256K1.java:81)\n\tat app//org.hyperledger.besu.crypto.SignatureAlgorithmType$$Lambda$201/0x000000080030b840.get(Unknown Source)\n\tat app//org.hyperledger.besu.crypto.SignatureAlgorithmFactory.getInstance(SignatureAlgorithmFactory.java:43)\n\tat app//org.hyperledger.besu.ethereum.core.Transaction.<init>(Transaction.java:98)\n\tat app//org.hyperledger.besu.ethereum.core.Transaction$Builder.build(Transaction.java:853)\n\tat app//org.hyperledger.besu.ethereum.core.encoding.TransactionDecoder.decodeFrontier(TransactionDecoder.java:126)\n\tat app//org.hyperledger.besu.ethereum.core.encoding.TransactionDecoder.decodeForWire(TransactionDecoder.java:66)\n\tat app//org.hyperledger.besu.ethereum.core.Transaction.readFrom(Transaction.java:105)\n\tat app//org.hyperledger.besu.ethereum.core.BlockBody$$Lambda$430/0x00000008004f8440.apply(Unknown Source)\n\tat app//org.hyperledger.besu.ethereum.rlp.RLPInput.readList(RLPInput.java:330)\n\tat app//org.hyperledger.besu.ethereum.core.BlockBody.readFrom(BlockBody.java:73)\n\tat app//org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStoragePrefixedKeyBlockchainStorage.lambda$getBlockBody$3(KeyValueStoragePrefixedKeyBlockchainStorage.java:84)\n\tat app//org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStoragePrefixedKeyBlockchainStorage$$Lambda$429/0x00000008004f8040.apply(Unknown Source)\n\tat java.base@11.0.10/java.util.Optional.map(Optional.java:265)\n\tat app//org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStoragePrefixedKeyBlockchainStorage.getBlockBody(KeyValueStoragePrefixedKeyBlockchainStorage.java:84)\n\tat app//org.hyperledger.besu.ethereum.chain.DefaultBlockchain.getBlockBody(DefaultBlockchain.java:196)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain.lambda$getBlockByHash$6(Blockchain.java:84)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain$$Lambda$479/0x0000000800523c40.apply(Unknown Source)\n\tat java.base@11.0.10/java.util.Optional.flatMap(Optional.java:294)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain.getBlockByHash(Blockchain.java:84)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain$$Lambda$1466/0x000000080080e840.apply(Unknown Source)\n\tat java.base@11.0.10/java.util.Optional.flatMap(Optional.java:294)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain.getBlockByNumber(Blockchain.java:88)\n\tat app//org.hyperledger.besu.ethereum.api.query.BlockchainQueries.lambda$gasPrice$28(BlockchainQueries.java:815)\n\tat app//org.hyperledger.besu.ethereum.api.query.BlockchainQueries$$Lambda$1462/0x000000080080f840.apply(Unknown Source)\n\tat java.base@11.0.10/java.util.stream.LongPipeline$1$1.accept(LongPipeline.java:177)\n\tat java.base@11.0.10/java.util.stream.Streams$RangeLongSpliterator.forEachRemaining(Streams.java:228)\n\tat java.base@11.0.10/java.util.Spliterator$OfLong.forEachRemaining(Spliterator.java:763)\n\tat java.base@11.0.10/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)\n\tat java.base@11.0.10/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)\n\tat java.base@11.0.10/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:550)\n\tat java.base@11.0.10/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)\n\tat java.base@11.0.10/java.util.stream.LongPipeline.toArray(LongPipeline.java:521)\n\tat app//org.hyperledger.besu.ethereum.api.query.BlockchainQueries.gasPrice(BlockchainQueries.java:823)\n\tat app//org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.EthGasPrice.response(EthGasPrice.java:54)\n\tat app//org.hyperledger.besu.ethereum.api.jsonrpc.JsonRpcHttpService.process(JsonRpcHttpService.java:724)\n\tat app//org.hyperledger.besu.ethereum.api.jsonrpc.JsonRpcHttpService.lambda$handleJsonSingleRequest$13(JsonRpcHttpService.java:578)\n\tat app//org.hyperledger.besu.ethereum.api.jsonrpc.JsonRpcHttpService$$Lambda$1414/0x00000008007e9c40.handle(Unknown Source)\n\tat app//io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)\n\tat app//io.vertx.core.impl.ContextImpl$$Lambda$1120/0x0000000800738840.run(Unknown Source)\n\tat java.base@11.0.10/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base@11.0.10/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base@11.0.10/java.lang.Thread.run(Thread.java:

**Comment by daniel-iobuilders** [comment]
@matkt Which command do you use to start the node? With `./gradlew run` I cannot reproduce the warning.

> When I start a node with the commit of this PR on master I have an issue
> 
> I tried with the commit just before and it's was ok
> 
> > {"timestamp":"2021-03-22T10:07:35,093","level":"WARN","thread":"vertx-blocked-thread-checker","class":"BlockedThreadChecker","message":"Thread Thread[vert.x-worker-thread-6,5,main]=Thread[vert.x-worker-thread-6,5,main] has been blocked for 62907 ms, time limit is 60000 ms","throwable":" io.vertx.core.VertxException: Thread blocked\n\tat java.base@11.0.10/java.util.HashMap.putVal(HashMap.java:637)\n\tat java.base@11.0.10/java.util.HashMap.put(HashMap.java:607)\n\tat java.base@11.0.10/java.security.Provider.implPut(Provider.java:1064)\n\tat java.base@11.0.10/java.security.Provider.put(Provider.java:511)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.addAlgorithm(BouncyCastleProvider.java:292)\n\tat app//org.bouncycastle.jcajce.provider.asymmetric.DH$Mappings.configure(Unknown Source)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.loadAlgorithms(BouncyCastleProvider.java:245)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.setup(BouncyCastleProvider.java:171)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.access$000(BouncyCastleProvider.java:58)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider$1.run(BouncyCastleProvider.java:153)\n\tat java.base@11.0.10/java.security.AccessController.doPrivileged(Native Method)\n\tat app//org.bouncycastle.jce.provider.BouncyCastleProvider.(BouncyCastleProvider.java:149)\n\tat app//org.hyperledger.besu.crypto.SECP256K1.(SECP256K1.java:81)\n\tat app//org.hyperledger.besu.crypto.SignatureAlgorithmType$$Lambda$201/0x000000080030b840.get(Unknown Source)\n\tat app//org.hyperledger.besu.crypto.SignatureAlgorithmFactory.getInstance(SignatureAlgorithmFactory.java:43)\n\tat app//org.hyperledger.besu.ethereum.core.Transaction.(Transaction.java:98)\n\tat app//org.hyperledger.besu.ethereum.core.Transaction$Builder.build(Transaction.java:853)\n\tat app//org.hyperledger.besu.ethereum.core.encoding.TransactionDecoder.decodeFrontier(TransactionDecoder.java:126)\n\tat app//org.hyperledger.besu.ethereum.core.encoding.TransactionDecoder.decodeForWire(TransactionDecoder.java:66)\n\tat app//org.hyperledger.besu.ethereum.core.Transaction.readFrom(Transaction.java:105)\n\tat app//org.hyperledger.besu.ethereum.core.BlockBody$$Lambda$430/0x00000008004f8440.apply(Unknown Source)\n\tat app//org.hyperledger.besu.ethereum.rlp.RLPInput.readList(RLPInput.java:330)\n\tat app//org.hyperledger.besu.ethereum.core.BlockBody.readFrom(BlockBody.java:73)\n\tat app//org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStoragePrefixedKeyBlockchainStorage.lambda$getBlockBody$3(KeyValueStoragePrefixedKeyBlockchainStorage.java:84)\n\tat app//org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStoragePrefixedKeyBlockchainStorage$$Lambda$429/0x00000008004f8040.apply(Unknown Source)\n\tat java.base@11.0.10/java.util.Optional.map(Optional.java:265)\n\tat app//org.hyperledger.besu.ethereum.storage.keyvalue.KeyValueStoragePrefixedKeyBlockchainStorage.getBlockBody(KeyValueStoragePrefixedKeyBlockchainStorage.java:84)\n\tat app//org.hyperledger.besu.ethereum.chain.DefaultBlockchain.getBlockBody(DefaultBlockchain.java:196)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain.lambda$getBlockByHash$6(Blockchain.java:84)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain$$Lambda$479/0x0000000800523c40.apply(Unknown Source)\n\tat java.base@11.0.10/java.util.Optional.flatMap(Optional.java:294)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain.getBlockByHash(Blockchain.java:84)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain$$Lambda$1466/0x000000080080e840.apply(Unknown Source)\n\tat java.base@11.0.10/java.util.Optional.flatMap(Optional.java:294)\n\tat app//org.hyperledger.besu.ethereum.chain.Blockchain.getBlockByNumber(Blockchain.java:88)\n\tat app//org.hyperledger.besu.ethereum.api.query.BlockchainQueries.lambda$gasPrice$28(BlockchainQueries.java:815)\n\tat app//org.hyperledger.besu.ethereum.api.query.BlockchainQueries$$Lambda$1462/0x000000080080f840.apply(Unknown Source)\n\tat java.base@11.0.10/java.util.stream.LongPipeline$1$1.accept(LongPipeline.java:177)\n\tat java.base@11.0.10/java.util.stream.Streams$RangeLongSpliterator.forEachRemaining(Streams.java:228)\n\tat java.base@11.0.10/java.util.Spliterator$OfLong.forEachRemaining(Spliterator.java:763)\n\tat java.base@11.0.10/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)\n\tat java.base@11.0.10/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)\n\tat java.base@11.0.10/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:550)\n\tat java.base@11.0.10/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260)\n\tat java.base@11.0.10/java.util.stream.LongPipeline.toArray(LongPipeline.java:521)\n\tat app//org.hyperledger.besu.ethereum.api.query.BlockchainQueries.gasPrice(BlockchainQueries.java:823)\n\tat app//org.hyperledger.besu.ethereum.api.jsonrpc.internal.methods.EthGasPrice.response(EthGasPrice.java:54)\n\tat app//org.hyperledger.besu.ethereum.api.jsonrpc.JsonRpcHttpService.process(JsonRpcHttpService.java:724)\n\tat app//org.hyperledger.besu.ethereum.api.jsonrpc.JsonRpcHttpService.lambda$handleJsonSingleRequest$13(JsonRpcHttpService.java:578)\n\tat app//org.hyperledger.besu.ethereum.api.jsonrpc.JsonRpcHttpService$$Lambda$1414/0x00000008007e9c40.handle(Unknown Source)\n\tat app//io.vertx.core.impl.ContextImpl.lambda$executeBlocking$2(ContextImpl.java:313)\n\tat app//io.vertx.core.impl.ContextImpl$$Lambda$1120/0x0000000800738840.run(Unknown Source)\n\tat java.base@11.0.10/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)\n\tat java.base@11.0.10/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)\n\tat app//io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)\n\tat java.base@11.0.10/java.lang.Thread.run(Thread.java:

**Comment by matkt** [comment]
> @matkt Which command do you use to start the node? With `./gradlew run` I cannot reproduce the warning.
> 
> > When I start a node with the commit of this PR on master I have an issue
> > I tried with the commit just before and it's was ok
> > > {"timestamp":"2021-03-22T10:07:35,093","level":"WARN","thread":"vertx-blocked-thread-checker","class":"BlockedThreadChecker","message":"Thread Thread[vert.x-worker-thread-6,5,main]=Thread[vert.x-worker-thread-6,5,main] has been blocked for 62907 ms, time limit is 60000 ms","throwable":" 

it is a node on aws (fastsync mainnet). It was already synchronized with a previous version. I have this error after the update during the startup 

Instance : i3.xlarge

config file 
```
network="mainnet"
min-gas-price=1000
sync-mode="FAST"
host-whitelist=["*"]
# rpc
rpc-http-enabled=true
rpc-http-host="0.0.0.0"
rpc-http-port=8545
rpc-http-api=["ADMIN","DEBUG","NET","ETH","MINER","WEB3"]
rpc-http-cors-origins=["all"]


# ws
rpc-ws-enabled=true
rpc-ws-api=["NET","ETH","WEB3"]
rpc-ws-host="0.0.0.0"
rpc-ws-port=8546

# graphql
graphql-http-enabled=true
graphql-http-host="0.0.0.0"
graphql-http-port=8547
graphql-http-cors-origins=["all"]

# p2p
p2p-host="18.222.219.219"
p2p-port=30303

# metrics
metrics-enabled=true
metrics-host="0.0.0.0"
metrics-port=9545
```

**Comment by lucassaldanha** [comment]
I might be wrong but from the stack trace it looks like we are trying to use Bouncy Castle when we should've been using the native secp255k1 implementation.And as a side effect it is blocking the Vertx thread. @daniel-iobuilders is investigating this.

**Comment by lucassaldanha** [comment]
We will revert this commit while we investigate.

**Pull request merged**